zeth_contracts/contracts/BLS12_377MixerBase.sol

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2015-2020 Clearmatics Technologies Ltd
+//
+// SPDX-License-Identifier: LGPL-3.0+
+
+pragma solidity ^0.5.0;
+pragma experimental ABIEncoderV2;
+
+import "./MixerBase.sol";
+import "./MiMC31.sol";
+
+// Partial implementation of abstract MixerBase which implements the
+// curve-specific methods to use the BLS12-377 pairing.
+contract BLS12_377MixerBase is MixerBase
+{
+    // TODO: Code here is very similar to AltBN128MixerBase, with only the
+    // constants changed. Look into sharing more code (possibly by making some
+    // of these constants dynamic).
+
+    // Constants regarding the hash digest length, the prime number used and
+    // its associated length in bits and the max values (v_in and v_out)
+    // field_capacity = floor( log_2(r) )
+    uint256 constant field_capacity = 252;
+
+    // Number of residual bits per bytes32
+    uint256 constant num_residual_bits = digest_length - field_capacity;
+
+    // Shift to move residual bits from lowest order to highest order
+    uint256 constant residual_bits_shift = 256 - num_residual_bits;
+
+    // Mask to extract the residual bits in the high-order position
+    uint256 constant residual_bits_mask =
+    ((1 << num_residual_bits) - 1) << residual_bits_shift;
+
+    // Total number of residual bits from packing of 256-bit long string into
+    // 253-bit long field elements to which are added the public value of size
```

Comment:
I was referring only to the "field elements" (i.e. using their "full" size). Their full size is not overly relevent relevant here, though, so removed this to avoid confusion.

Reply:
I see, thanks